### PR TITLE
HPCC-22827 Add employeeNumber to wsAccess

### PR DESCRIPTION
**Pull request by RussWhitehead**
Add an employeeNumber field to wsAccess, ISECUser and ldapconnection.

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@wangkx Please review

File: esp/services/ws_access/ws_accessService.cpp [hunk 1]
```diff
@@ -941,6 +953,11 @@ bool Cws_accessEx::onAddUser(IEspContext &context, IEspAddUserRequest &req, IEsp
             employeeID = req.getEmployeeID();
         }
 
+        const char * employeeNumber = NULL;
```
Comment (wangkx) [inline esp/services/ws_access/ws_accessService.cpp]: Minor: should use 'nullptr'

File: esp/services/ws_access/ws_accessService.cpp [hunk 2]
```diff
@@ -951,6 +968,8 @@ bool Cws_accessEx::onAddUser(IEspContext &context, IEspAddUserRequest &req, IEsp
             user->setLastName(lastname);
         if(employeeID != NULL)
             user->setEmployeeID(employeeID);
+        if(employeeNumber != NULL)
```
Comment (wangkx) [inline esp/services/ws_access/ws_accessService.cpp]: same as the above

File: esp/services/ws_access/ws_accessService.cpp [hunk 3]
```diff
@@ -3709,6 +3735,11 @@ bool Cws_accessEx::onUserInfoEdit(IEspContext &context, IEspUserInfoEditRequest
             user->setEmployeeID(req.getEmployeeID());
         }
 
+        if (context.getClientVersion() >= 1.15)
+        {
+            user->setEmployeeNumber(req.getEmployeeNumber());
```
Comment (wangkx) [inline esp/services/ws_access/ws_accessService.cpp]: Check if req.getEmployeeNumber() is nullptr?

File: esp/services/ws_access/ws_accessService.cpp [hunk 4]
```diff
@@ -3766,7 +3797,12 @@ bool Cws_accessEx::onUserInfoEditInput(IEspContext &context, IEspUserInfoEditInp
                 StringBuffer sb;
                 resp.setPasswordExpiration(getPasswordExpiration(user, sb));
             }
+            if (version >= 1.15)
+            {
+                resp.setEmployeeNumber(user->getEmployeeNumber());
```
Comment (wangkx) [inline esp/services/ws_access/ws_accessService.cpp]: Check if user->getEmployeeNumber() is nullptr?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 5]
```diff
@@ -2694,6 +2708,7 @@ class CLdapClient : implements ILdapClient, public CInterface
         userTree->addPropInt(getUserFieldNames(UFUserID), usr.getUserID());
         userTree->addProp(getUserFieldNames(UFPasswordExpiration), sb.str());
         userTree->addProp(getUserFieldNames(UFEmployeeID), usr.getEmployeeID());
+        userTree->addProp(getUserFieldNames(UFEmployeeNumber), usr.getEmployeeNumber());
```
Comment (wangkx) [inline system/security/LdapSecurity/ldapconnection.cpp]: Check if usr.getEmployeeNumber() is nullptr?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 6]
```diff
@@ -6044,7 +6070,15 @@ class CLdapClient : implements ILdapClient, public CInterface
             employeeID_values
         };
 
-        LDAPMod *attrs[9];
+        char* employeeNumber_values[] = {(char*)employeeNumber, NULL};
```
Comment (wangkx) [inline system/security/LdapSecurity/ldapconnection.cpp]: Check if isEmptyString(employeeNumber)?

File: system/security/LdapSecurity/ldapsecurity.cpp [hunk 7]
```diff
@@ -250,6 +261,7 @@ void CLdapSecUser::copyTo(ISecUser& destination)
     dest->setFirstName(getFirstName());
     dest->setLastName(getLastName());
     dest->setEmployeeID(getEmployeeID());
+    dest->setEmployeeNumber(getEmployeeNumber());
```
Comment (wangkx) [inline system/security/LdapSecurity/ldapsecurity.cpp]: Check if nullptr?

File: system/security/shared/SecureUser.hpp [hunk 8]
```diff
@@ -250,6 +262,7 @@ class CSecureUser : implements ISecUser, implements ISecCredentials, public CInt
         destination.setFirstName(getFirstName());
         destination.setLastName(getLastName());
         destination.setEmployeeID(getEmployeeID());
+        destination.setEmployeeNumber(getEmployeeNumber());
```
Comment (wangkx) [inline system/security/shared/SecureUser.hpp]: Check if nullptr?

File: esp/services/ws_access/ws_accessService.cpp [hunk 9]
```diff
@@ -497,6 +497,11 @@ bool Cws_accessEx::onUsers(IEspContext &context, IEspUserRequest &req, IEspUserR
                     StringBuffer sb;
                     oneusr->setPasswordexpiration(getPasswordExpiration(usr, sb));
                 }
+
+                if (version >= 1.15)
+                {
+                    oneusr->setEmployeeNumber(usr->getEmployeeNumber());
```
Comment (wangkx) [inline esp/services/ws_access/ws_accessService.cpp]: Check if usr->getEmployeeNumber() is nullptr?

File: esp/services/ws_access/ws_accessService.cpp [hunk 10]
```diff
@@ -1438,6 +1460,10 @@ bool Cws_accessEx::onGroupMemberQuery(IEspContext &context, IEspGroupMemberQuery
             if (employeeID && *employeeID)
                 userInfo->setEmployeeID(employeeID);
 
+            const char* employeeNumber = usr.queryProp(getUserFieldNames(UFEmployeeNumber));
+            if (employeeNumber && *employeeNumber)
+                userInfo->setEmployeeNumber(employeeNumber);
```
Comment (wangkx) [inline esp/services/ws_access/ws_accessService.cpp]: Check client version?

**Review by wangkx — COMMENTED** [review]
@RussWhitehead  minors

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 6 above)
Comment (RussWhitehead) [reply system/security/LdapSecurity/ldapconnection.cpp]: Have to declare the struct whether  or not employeeNumber us empty (struct is referenced below)

**Review by wangkx — APPROVED** [review]
@RussWhitehead looks fine.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Comment by richardkchapman** [comment]
@RussWhitehead Is 7.6.0 a more appropriate target than 7.6.2 ?

**Comment by RussWhitehead** [comment]
@richardkchapman No pressing need for this enhancement, but if you are okay with 7.6.0 that would be fine. Shall I rebase?

**Comment by richardkchapman** [comment]
@RussWhitehead Please

**Comment by richardkchapman** [comment]
(and squash)

**Pull request closed without merging**